aws-lc-rs/build.rs

```diff
@@ -2,28 +2,74 @@
 // SPDX-License-Identifier: Apache-2.0 OR ISC
 
 fn main() {
-    let mutually_exclusives_count = [cfg!(feature = "non-fips"), cfg!(feature = "fips")]
-        .iter()
-        .filter(|x| **x)
-        .count();
-
-    if mutually_exclusives_count > 1 {
-        eprint!("fips and non-fips are mutually exclusive crate features.");
-        std::process::exit(1);
-    }
+    let has_mutually_exclusive_features = cfg!(feature = "non-fips") && cfg!(feature = "fips");
+    assert!(
+        !has_mutually_exclusive_features,
+        "`fips` and `non-fips` are mutually exclusive crate features."
+    );
 
     // This appears asymmetric, but it reflects the `cfg` statements in lib.rs that
     // require `aws-lc-sys` to be present when "fips" is not enabled.
-    let at_least_one_count = [cfg!(feature = "aws-lc-sys"), cfg!(feature = "fips")]
-        .iter()
-        .filter(|x| **x)
-        .count();
-
-    if at_least_one_count < 1 {
-        eprint!(
-            "one of the following features must be specified: \
-        aws-lc-sys, non-fips, or fips."
+    // if `fips` is enabled, then use that
+    let sys_crate = if cfg!(feature = "fips") {
+        "aws-lc-fips-sys"
+    } else if cfg!(feature = "aws-lc-sys") {
+        "aws-lc-sys"
+    } else {
+        panic!(
+            "one of the following features must be specified: `aws-lc-sys`, `non-fips`, or `fips`."
         );
-        std::process::exit(1);
+    };
+
+    export_sys_vars(sys_crate);
+}
+
+fn export_sys_vars(sys_crate: &str) {
+    let prefix = if sys_crate == "aws-lc-fips-sys" {
+        "DEP_AWS_LC_FIPS_"
+    } else {
+        "DEP_AWS_LC_"
+    };
+
+    let mut selected = String::default();
+    let mut candidates = vec![];
+
+    // search through the DEP vars and find the selected sys crate version
+    for (name, value) in std::env::vars() {
```

Comment:
I know it's less efficient, but I think it would be easier to read if it would first identify the selected version, then try to export the variables.
```
    let mut selected = String::default();
    // search through the DEP vars and find the selected sys crate version
    for (name, _) in std::env::vars() {
        // we're still looking for a selected prefix
        if let Some(version) = name.strip_prefix(prefix) {
            if let Some(version) = version.strip_suffix("_INCLUDE") {
                // we've found the selected version so update it
                selected = format!("{prefix}{version}_");
                break;
            }
        }
    }
    assert!(!selected.is_empty(), "missing {prefix} include");

    // process all environment vars
    for (name, value) in std::env::vars() {
        if let Some(var) = name.strip_prefix(&selected) {
            eprintln!("cargo:rerun-if-env-changed={name}");
            let var = var.to_lowercase();
            println!("cargo:{var}={value}");
        }
    }
```

Reply:
that's actually _literally_ what i had initially. but i do worry that we're iterating over this twice, which may be quite large in some environments and make build times quite a bit worse. what i ended up with _should_ do it in a single pass (assuming it's alphabetical), which seems not nearly as bad.